### DISCUSSION TITLE
how to test button which is invisible and  mock router.isReady in test?

### DISCUSSION THREAD

**Discussion by LuYujin9** [neuefische/questions#553] [Web] [answered]
Hi, I am struggling with testing and have 3 problems. Could you help me?

 **in Album.test.js**
Problem 1 : how to test "doesn't render toNextPhotoButton when it shows the last photo
Problem 2 : how to test "be called time", when the handle function is not a props.

 **in Heading.test.js**
Problem3: 
I use 
` if (!router.isReady) {
    return <h1>loading</h1>;
  }`to make sure the other processes run after getting "router". But in the test , it renders only `<h1>loading<h1>` but not `<h1>title<h1>` which I expected.

Here ist my code:

https://github.com/LuYujin9/capstone-project/pull/7/files

**Comment by PReichetanz** [comment, marked as answer]
Hi @LuYujin9,

really nice tests! I've found a way to solve your problem.

## Problem 1: how to test "doesn't render toNextPhotoButton when it shows the last photo
In your current code, this is not a problem: the test works and checks that the `toNextPhotoButton` is not present when clicking the button three times.

## Problem 2: how to test "be called time", when the handle function is not a props.
It's not possible and we should do that: we would test implementation details which we never should do.
So you need to remove the last to `expect(...).toHaveBeenCalledTimes()` lines.

Besides: the matcher was not correct, it is `toHaveBeenCalledTimes()` instead of `toBeCalledTimes()` (as far as I know).


## Problem 3: Heading.test.js
Simply mock the `isReady` value and set it to true:
```js
jest.mock('next/router', () => ({
  useRouter() {
    return {
      push: jest.fn(),
      asPath: '/',
      isReady: true,
    };
  },
}));
```

**Reply by LuYujin9** [reply]
Thanks, Paul. I have solved most of problems with your suggestions. but  the problem 2 in Album Test not:
 I use "spyOn" to spy the  "handleToNextPhoto" function in component "Album"
`const handleToNextPhoto = jest.spyOn(Album, "handleToNextPhoto");`
in Album, it does a function
` function handleToNextPhoto() {
    photoIndex < photos.length - 1 ? setPhotoIndex(photoIndex + 1) : photoIndex;
  }` 
but it showed this in test: "Cannot spy on the handleToNextPhoto property because it is not a function "

I have pushed my new code. There ist the link: 
https://github.com/LuYujin9/capstone-project/pull/7/files#diff-f87f123ae9a1d6714a7be5252d970e344f26e8d7ee73746ac55beb5eb3553fb0

**Reply by PReichetanz** [reply]
@LuYujin9  Yes, because we don't test implementation details with component testing: we should only test functions which are passed as props to a component.

I've found [this StackOverflow conversation](https://stackoverflow.com/questions/69306919/spy-on-a-function-inside-functional-component-using-react-testing-library) where the answers point into the same direction.

I would recommend to not test this function; in a production environment, this would be covered by end-to-end-testing of a QA expert.

**Reply by LuYujin9** [reply]
I thought it too complex. Thank you. Finally I can post my PR again. :)
